app/src/main/java/de/mathfactory/mooltifill/MooltipassDevice.kt

```diff
@@ -428,18 +430,71 @@ class MooltipassDevice(private val device: BluetoothDevice, private var debug: I
 @ExperimentalCoroutinesApi
 class MooltipassScan {
 
+    private var context:Context? = null
+
     private fun pairedDevice(context: Context): BluetoothDevice? {
+        scheduleBlutoothStateObserver(context);
+
         val bluetoothManager = context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
-        return bluetoothManager.adapter.bondedDevices.lastOrNull(::filter)
+        return bluetoothManager.adapter.bondedDevices.filter(::filter).firstOrNull()
     }
 
     fun deviceFlow(context: Context): Flow<BluetoothDevice> {
         return pairedDevice(context)
-            ?.let(::flowOf) // use paired device, if available...
-            ?: emptyFlow() // do not scan, as paired device is necessary
+                ?.let(::flowOf) // use paired device, if available...
+                ?: emptyFlow() // do not scan, as paired device is necessary
 //            ?:scanFlow(context).map(ScanResult::getDevice) // ... else scan devices
     }
 
+    private fun scheduleBlutoothStateObserver(context: Context)
+    {
+        this.context = context
+        val filter = IntentFilter()
+        filter.addAction(BluetoothDevice.ACTION_ACL_CONNECTED)
+        context.registerReceiver(broadcastReceiver, filter)
```

Comment:
You are using Context.registerReceiver() here but I do not see a corresponding Context.unregisterReceiver(), which will leak memory... and the receiver will be registered multiple times on each invocation of pairedDevice(), I think? Also, regarding the registering the receiver through the pairedDevice() function: Did you verify this solution works when e.g. not using the "Mooltipass Awareness" option? I think pairedDevice() in this case will get called not until the instance when device access is made, which might be too late to get the ACTION_ACL_CONNTECTED event.